python/mead/downloader.py

```diff
@@ -0,0 +1,300 @@
+        super(SingleFileDownloader, self).__init__(data_download_cache, cache_ignore)
+        self.dataset_file = dataset_file
+        self.data_download_cache = data_download_cache
+
+    def download(self):
+        import json
+        file_loc = self.dataset_file
+        if check_sanity_file(file_loc):
+            return file_loc
+        elif validate_url(file_loc):  # is it a web URL? check if exists in cache
+            url = file_loc
+            try:
+                dcaches = json.load(open(os.path.join(self.data_download_cache, "datasets-embeddings-cache.json")))
+            except IOError:
+                dcaches = {}
+            if url in dcaches and check_sanity_file(dcaches[url], url) and not self.cache_ignore:
+                print("file for {} found in cache, not downloading".format(url))
+                return dcaches[url]
+            else:  # download the file in the cache, update the json
+                cache_dir = self.data_download_cache
+                print("using {} as data/embeddings cache".format(cache_dir))
+                zipd = {'application/gzip': extract_gzip, 'application/zip': extract_zip}
+                temp_file = web_downloader(url)
+                dload_file = extractor(filepath=temp_file, cache_dir=cache_dir, extractor_func=zipd.get(mime_type(temp_file), None))
+                dcaches.update({url: dload_file})
+                json.dump(dcaches, open(os.path.join(self.data_download_cache, "datasets-embeddings-cache.json"), "w"), indent=True)
+                return dload_file
+        else:
+            raise RuntimeError("the file {} is not in cache and can not be downloaded".format(file_loc))
+
+
+class DataDownloader(Downloader):
+    def __init__(self, dataset_desc, data_download_cache, enc_dec=False, cache_ignore=False):
+        super(DataDownloader, self).__init__(data_download_cache, cache_ignore)
+        self.dataset_desc = dataset_desc
+        self.data_download_cache = data_download_cache
+        self.enc_dec = enc_dec
+
+    def download(self):
+        import os
+        import json
+        dload_bundle = self.dataset_desc.get("download", None)
+        if dload_bundle is not None:  # download a zip/tar/tar.gz directory, look for train, dev test files inside that.
+            try:
+                dcaches = json.load(open(os.path.join(self.data_download_cache, "datasets-embeddings-cache.json")))
+            except IOError:
+                dcaches = {}
+            if dload_bundle in dcaches and check_sanity_dir(dcaches[dload_bundle], self.dataset_desc, self.enc_dec, dload_bundle) and not self.cache_ignore:
+                download_dir = dcaches[dload_bundle]
+                print("files for {} found in cache, not downloading".format(dload_bundle))
+                return {k: os.path.join(download_dir, self.dataset_desc[k]) for k in self.dataset_desc if k.endswith("_file")}
+            else:  # try to download the bundle and unzip
+                url = dload_bundle
```

Comment:
this feels like an unnecessary indirection. If url is more descriptive of the value, then change the name of dload_bundle.